tests/functional_tests/util_resources.py

```diff
@@ -0,0 +1,117 @@
+# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+# THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+# THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""
+    Help determine how much CPU power is available at the given time
+    by running numerical calculations
+"""
+
+from __future__ import print_function
+import datetime
+import multiprocessing
+import psutil
+from math import pi
+
+def available_ram_gb():
+    ram_bytes = psutil.virtual_memory().available
+    kilo = 1024
+    ram_gb = ram_bytes / kilo / kilo / kilo
+    return ram_gb
+
+def pi_digits_lib(digits):
+    """
+    Returns the number of digits of Pi
+    """
+    return float('{:.{}f}'.format(pi, digits))
+
+def pi_digits_numeric_yield(digits):
+    """Generate x digits of Pi.
+    https://stackoverflow.com/a/54967370
+    """
+    # Make the format the same as the library call
+    # here we need 2 chars more for '3' and '.'
+    x = digits + 2
+    k,a,b,a1,b1 = 2,4,1,12,4
+    while x > 0:
+        p,q,k = k * k, 2 * k + 1, k + 1
+        a,b,a1,b1 = a1, b1, p*a + q*a1, p*b + q*b1
+        d,d1 = a/b, a1/b1
+        while d == d1 and x > 0:
+            yield int(d)
+            x -= 1
+            a,a1 = 10*(a % b), 10*(a1 % b1)
+            d,d1 = a/b, a1/b1
+
+
+def pi_digits_numeric(digits):
+    "Get PI rounded to n digits"
+    my_array = []
+    for i in pi_digits_numeric_yield(digits):
+        my_array.append(str(i))
+        
+    first = my_array.pop(0) # Pop the 1st 3
+    valStr = "{}.{}".format(first, "".join(my_array))
+    val = float(valStr)
+    rounded = round(val, digits)
+    
+    # Compare if the calcualted Pi is the same as the library one
+    pi_library = pi_digits_lib(digits)
+    if pi_library != rounded:
+        raise RuntimeError("Calculated Pi = ", rounded, "Library Pi = ", pi_library)
+    return rounded
+        
+def get_time_pi(num_digits, cores=multiprocessing.cpu_count()):
+    """   
+    Get the time of generation of pi, using the chosen number of cores.
+    This helps in estimating how well the problem scales on multiple cores.
+    """
+    # Inputs are the number of digits to calculate for each CPU
+    inp = [num_digits] * cores
+    # Start measuring the time
+    start = datetime.datetime.now()
+    # Perform the calculations on multiple cores in parallel to stress the CPU
+    #with multiprocessing.Pool(cores) as pool: # Python 3.3+
+    #    pool.map(pi_digits_numeric, inp)
+    pool = multiprocessing.Pool(cores)  # Python < 3.3
```

Comment:
Should probably use a try/finally here (terminate in the finally).

Reply:
I'm not sure about this. If I catch the exception, then I should handle it with a backup solution. For this test an exception here would have to mean aborting the test.

Reply:
try/finally doesn't swallow any exceptions.

Reply:
OK then.

Reply:
Not needed anymore. Launching an external C++ app.